File: app/src/main/java/tech/salroid/filmy/CustomAdapter/SavedMoviesAdapter.java
package tech.salroid.filmy.CustomAdapter;

/**
 * Created by Home on 7/30/2016.
 */

import android.content.Context;
import android.database.Cursor;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;

import tech.salroid.filmy.Database.FilmContract;
import tech.salroid.filmy.R;



/**
 * Created by Home on 7/20/2016.
 */
public class SavedMoviesAdapter extends RecyclerView.Adapter<SavedMoviesAdapter.Vh> {

    private final LayoutInflater inflater;
    Context context;
    private ClickListener clickListener;
    private Cursor dataCursor;
    private LongClickListener longclickListener;

    public SavedMoviesAdapter(Context context, Cursor cursor) {

        inflater = LayoutInflater.from(context);
        this.context = context;
        this.dataCursor = cursor;

    }


    @Override
    public Vh onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = inflater.inflate(R.layout.custom_row, parent, false);
        Vh holder = new Vh(view);

        return holder;
    }

    @Override
    public void onBindViewHolder(Vh holder, int position) {

        String movie_title, imdb_id, movie_poster;
        int movie_year;

        dataCursor.moveToPosition(position);

        int id_index = dataCursor.getColumnIndex(FilmContract.SaveEntry.SAVE_ID);
        int title_index = dataCursor.getColumnIndex(FilmContract.SaveEntry.SAVE_TITLE);
        int poster_index = dataCursor.getColumnIndex(FilmContract.SaveEntry.SAVE_POSTER_LINK);
        int year_index = dataCursor.getColumnIndex(FilmContract.SaveEntry.SAVE_YEAR);


        movie_title = dataCursor.getString(title_index);
        movie_poster = dataCursor.getString(poster_index);
        imdb_id = dataCursor.getString(id_index);
        movie_year = dataCursor.getInt(year_index);


        holder.title.setText(movie_title);
        Glide.with(context).load(movie_poster).into(holder.poster);


    }


    @Override
    public int getItemCount() {
        return (dataCursor == null) ? 0 : dataCursor.getCount();
    }

    public Cursor swapCursor(Cursor cursor) {
        if (dataCursor == cursor) {
            return null;
        }
        Cursor oldCursor = dataCursor;
        this.dataCursor = cursor;
        if (cursor != null) {
            this.notifyDataSetChanged();
        }
        return oldCursor;
    }


    class Vh extends RecyclerView.ViewHolder {

        TextView title;
        ImageView poster;
        FrameLayout main;

        public Vh(View itemView) {

            super(itemView);

            title = (TextView) itemView.findViewById(R.id.title);
            poster = (ImageView) itemView.findViewById(R.id.poster);
            main = (FrameLayout) itemView.findViewById(R.id.main);


            main.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {

                    dataCursor.moveToPosition(getPosition());

                    int movie_id_index = dataCursor.getColumnIndex(FilmContract.SaveEntry.SAVE_ID);
                    int movie_title_index = dataCursor.getColumnIndex(FilmContract.SaveEntry.SAVE_TITLE);

                    if (clickListener != null) {
                        clickListener.itemClicked(dataCursor.getString(movie_id_index),dataCursor.getString(movie_title_index));
                    }

                }
            });

            main.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View view) {

                    dataCursor.moveToPosition(getPosition());


                    if (longclickListener != null) {
                        longclickListener.itemLongClicked(dataCursor,getPosition());
                    }

                    return true;
                }
            });

            }
        }



    public interface ClickListener {

        public void itemClicked(String id,String title);

    }

    public interface LongClickListener{
        public void itemLongClicked(Cursor cursor,int position);

    }

    public void setLongClickListener(LongClickListener longclickListener) {
        this.longclickListener = longclickListener;
    }

    public void setClickListener(ClickListener clickListener) {
        this.clickListener = clickListener;
    }

}

File: app/src/main/java/tech/salroid/filmy/ui/adapters/SavedMoviesAdapter.kt
package tech.salroid.filmy.ui.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import tech.salroid.filmy.data.local.db.entity.MovieDetails
import tech.salroid.filmy.databinding.CustomRowBinding
import tech.salroid.filmy.utility.toReadableDate

class SavedMoviesAdapter(
    private var moviesList: ArrayList<MovieDetails>? = null,
    private val clickListener: ((String, String?) -> Unit)? = null,
    private val longClickListener: ((MovieDetails, Int) -> Unit)? = null
) : RecyclerView.Adapter<SavedMoviesAdapter.SavedMoviesViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SavedMoviesViewHolder {
        val binding =
            CustomRowBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return SavedMoviesViewHolder(binding)
    }

    override fun onBindViewHolder(holder: SavedMoviesViewHolder, position: Int) {
        moviesList?.let { holder.bindData(it[position]) }
    }

    override fun getItemCount(): Int = moviesList?.size ?: 0

    fun swapData(moviesList: ArrayList<MovieDetails>) {
        this.moviesList = moviesList
        notifyDataSetChanged()
    }

    fun removeItemAt(position: Int){
        moviesList?.removeAt(position)
        notifyItemRemoved(position)
    }

    inner class SavedMoviesViewHolder(private val binding: CustomRowBinding) :
        RecyclerView.ViewHolder(binding.root) {

        fun bindData(movie: MovieDetails) {
            binding.movieName.text = movie.title
            binding.movieYear.text = movie.releaseDate?.toReadableDate()

            Glide.with(binding.root.context)
                .load("http://image.tmdb.org/t/p/w342${movie.posterPath}")
                .into(binding.poster)
        }

        init {
            binding.main.setOnClickListener {
                moviesList?.let {
                    clickListener?.invoke(
                        it[adapterPosition].id.toString(),
                        it[adapterPosition].originalTitle
                    )
                }
            }

            binding.main.setOnLongClickListener {
                moviesList?.let { it1 ->
                    longClickListener?.invoke(
                        it1[adapterPosition],
                        adapterPosition
                    )
                }
                true
            }
        }
    }
}
